py/kubeflow/kfctl/testing/util/gcp_util.py

```diff
@@ -122,46 +122,84 @@ def iap_is_ready(url, wait_min=15):
     sleep(10)
   return False
 
-def basic_auth_is_ready(url, username, password, wait_min=15):
-  get_url = url + "/kflogin"
-  post_url = url + "/apikflogin"
+def _send_req(wait_sec, url, req_gen):
+  """ Helper function to send requests and retry when the endpoint is not ready.
 
-  req_num = 0
-  end_time = datetime.datetime.now() + datetime.timedelta(
-      minutes=wait_min)
-  while datetime.datetime.now() < end_time:
-    req_num += 1
-    logging.info("Trying url: %s request number %s" % (get_url, req_num))
+  Args:
+  wait_sec: int, max time to wait and retry in seconds.
+  url: str, url to send the request, used only for logging.
+  req_gen: lambda, no parameter function to generate requests.Request for the
+  function to send to the endpoint.
+
+  Returns:
+    requests.Response
+  """
+
+  # Simple identifier that we need to re-send the request.
+  class RetryError(Exception):
+    pass
+
+  def retry_if_retry_error(e):
+    return isinstance(e, RetryError)
+
+  @retry(stop_max_delay=wait_sec * 1000, wait_fixed=10 * 1000,
+         retry_on_exception=retry_if_retry_error)
+  def _send(url, req_gen):
     resp = None
+    logging.info("sending request to %s" % url)
     try:
-      resp = requests.request(
-          "GET",
-          get_url,
-          verify=False)
+      resp = req_gen()
     except SSLError as e:
-      logging.warning("%s: Endpoint SSL handshake error: %s; request number: %s" % (url, e, req_num))
+      logging.warning("%s: Endpoint SSL handshake error: %s" % (url, e))
+      raise RetryError()
     except ReqConnectionError:
       logging.info(
-          "%s: Endpoint not ready, request number: %s" % (url, req_num))
+          "%s: Endpoint not ready" % url)
+      raise RetryError()
+    except Exception as e:
+      logging.info("%s: unexpected error: %s" % (url, e))
+      raise e
+    return resp
+
+  return _send(url, req_gen)
+
+def basic_auth_is_ready(url, username, password, wait_min=15):
+  get_url = url + "/kflogin"
+  post_url = url + "/apikflogin"
+
+  end_time = datetime.datetime.now() + datetime.timedelta(
+      minutes=wait_min)
+
+  wait_time = datetime.datetime.now() - end_time
+  while True:
+    resp = _send_req(wait_time.seconds, get_url, lambda: requests.request(
```

Comment:
nit: I think the code would be simpler if you just did something like

```

 def retry_if_retry_error(e):
    return isinstance(e, [SSLError, ReqConnectionError])

@retry(..., retry_on_exception=retry_if_retry_error)
def send_request(*args,**kwargs):
  return requests.request(*args, **kwargs)
```

Reply:
so we need to be flexible about the condition to retry:
- some exceptions (SSLError, ReqConnectionError)
- some status code (200)

so I added an optional param to `_send_req` for the purpose.